drivers/gles3/storage/mesh_storage.cpp

```diff
@@ -2085,60 +2142,114 @@ void MeshStorage::_update_dirty_multimeshes() {
+		}
 
-			if (multimesh->data_cache_used_dirty_regions) {
-				uint32_t data_cache_dirty_region_count = Math::division_round_up(multimesh->instances, (int)MULTIMESH_DIRTY_REGION_SIZE);
-				uint32_t visible_region_count = visible_instances == 0 ? 0 : Math::division_round_up(visible_instances, (uint32_t)MULTIMESH_DIRTY_REGION_SIZE);
+		_update_dirty_multimesh(multimesh, uses_motion_vectors);
 
-				GLint region_size = multimesh->stride_cache * MULTIMESH_DIRTY_REGION_SIZE * sizeof(float);
+		multimesh_dirty_list = multimesh->dirty_list;
 
-				if (multimesh->data_cache_used_dirty_regions > 32 || multimesh->data_cache_used_dirty_regions > visible_region_count / 2) {
-					// If there too many dirty regions, or represent the majority of regions, just copy all, else transfer cost piles up too much
-					glBindBuffer(GL_ARRAY_BUFFER, multimesh->buffer);
-					glBufferSubData(GL_ARRAY_BUFFER, 0, MIN(visible_region_count * region_size, multimesh->instances * multimesh->stride_cache * sizeof(float)), data);
-					glBindBuffer(GL_ARRAY_BUFFER, 0);
-				} else {
-					// Not that many regions? update them all
-					// TODO: profile the performance cost on low end
-					glBindBuffer(GL_ARRAY_BUFFER, multimesh->buffer);
-					for (uint32_t i = 0; i < visible_region_count; i++) {
-						if (multimesh->data_cache_dirty_regions[i]) {
-							GLint offset = i * region_size;
-							GLint size = multimesh->stride_cache * (uint32_t)multimesh->instances * (uint32_t)sizeof(float);
-							uint32_t region_start_index = multimesh->stride_cache * MULTIMESH_DIRTY_REGION_SIZE * i;
-							glBufferSubData(GL_ARRAY_BUFFER, offset, MIN(region_size, size - offset), &data[region_start_index]);
-						}
-					}
-					glBindBuffer(GL_ARRAY_BUFFER, 0);
-				}
+		multimesh->dirty_list = nullptr;
+		multimesh->dirty = false;
+	}
+
+	multimesh_dirty_list = nullptr;
+}
+
+void MeshStorage::_update_dirty_multimesh(MultiMesh *p_multimesh, bool p_uses_motion_vectors) {
+	if (p_multimesh->data_cache.size()) { // May have been cleared, so only process if it exists.
+		const float *data = p_multimesh->data_cache.ptr();
 
-				for (uint32_t i = 0; i < data_cache_dirty_region_count; i++) {
-					multimesh->data_cache_dirty_regions[i] = false;
+		uint32_t visible_instances = p_multimesh->visible_instances >= 0 ? p_multimesh->visible_instances : p_multimesh->instances;
+
+		if (p_multimesh->data_cache_used_dirty_regions) {
+			uint32_t data_cache_dirty_region_count = Math::division_round_up(p_multimesh->instances, (int)MULTIMESH_DIRTY_REGION_SIZE);
+			uint32_t visible_region_count = visible_instances == 0 ? 0 : Math::division_round_up(visible_instances, (uint32_t)MULTIMESH_DIRTY_REGION_SIZE);
+
+			GLint region_size = p_multimesh->stride_cache * MULTIMESH_DIRTY_REGION_SIZE * sizeof(float);
+
+			if ((p_multimesh->data_cache_used_dirty_regions > 32 || p_multimesh->data_cache_used_dirty_regions > visible_region_count / 2) || p_uses_motion_vectors) {
```

Comment:
Super minor thing, but this can be slightly simplified by removing the inner parens, like:

```suggestion
			if (p_multimesh->data_cache_used_dirty_regions > 32 || p_multimesh->data_cache_used_dirty_regions > visible_region_count / 2 || p_uses_motion_vectors) {
```